internals/src/error.rs

```diff
@@ -26,3 +26,23 @@ macro_rules! write_err {
         }
     }
 }
+
+/// Trait used internally to assert that error types implement `Send` and `Sync`.
+#[doc(hidden)]
+pub trait AssertSendSync: Send + Sync {}
+
+/// Trait used internally to assert that error types implement `Debug`.
+#[doc(hidden)]
+pub trait AssertDebug: core::fmt::Debug {}
+
+/// Checks at compile time that type `$t` implements the traits we require for error types.
+#[macro_export]
+macro_rules! check_pub_error {
+    ($t:ty) => {
+        #[cfg(test)]
```

Comment:
It just occurred to me that I may be wrong and this doesn't work either. I would've checked but need to go rn, so I suggest you check it (by commenting-out an impl) when addressing my other comment.